### PR TITLE
feature/oauth-firebase

### PR DESCRIPTION
**Pull request by aish-26-2000**

### CODE REVIEW

File: src/modules/oauth/oauth.service.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import config from '../../config/config';
+import { logger } from '../logger';
+import { firebaseAdmin } from './firebase';
+
+/**
+ * This function generates options for a cookie based on the config settings.
+ *
+ * @returns {Promise<Object>} - A promise that resolves to an options object for a cookie
+ */
+export const generateCookieOptions = async () => {
```
Comment (sonalidevbud) [inline src/modules/oauth/oauth.service.ts:10]: please remove this is not used anywhere

File: src/config/config.ts [hunk 2]
```diff
@@ -65,6 +65,25 @@ const config = {
     from: envVars.EMAIL_FROM,
   },
   clientUrl: envVars.CLIENT_URL,
+  firebase: {
+    apiKey: envVars.API_KEY,
+    clientEmail: envVars.CLIENT_EMAIL,
+    privateKey: envVars.PRIVATE_KEY,
+    projectId: envVars.PROJECT_ID,
+  },
+  google: {
```
Comment (sonalidevbud) [inline src/config/config.ts:74]: please remove google and github creds if they are not getting used anywhere

File: src/modules/oauth/oauth.middleware.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+ * If the token is valid and the user has the required rights, the function adds the user object
+ *
+ * @param requiredRights - An array of strings representing the required rights for the user to access the route.
+ * @returns An async middleware function that takes in the request, response, and next middleware function.
+ * @throws ApiError with status code 401 (Unauthorized) if the token is missing or invalid.
+ * @throws ApiError with status code 403 (Forbidden) if the user does not have the required rights.
+ * @throws Other errors thrown by the Firebase Admin SDK functions.
+ */
+// eslint-disable-next-line import/prefer-default-export
+export const Authenticate =
```
Comment (sonalidevbud) [inline src/modules/oauth/oauth.middleware.ts:31]: change name to isAuthenticated

File: src/modules/oauth/oauth.middleware.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    // If there's no token, return an error
+    if (!token) {
+      return next(new ApiError(httpStatus.UNAUTHORIZED, 'Please authenticate'));
+    }
+    try {
+      // Verify the ID token using Firebase Admin SDK
+      const decodedToken = await firebaseAdmin.auth().verifyIdToken(token);
+
+      // Retrieve the user data from the database using the Firebase UID
+      const user: any = await getUserByFirebaseId(decodedToken.uid);
```
Comment (sonalidevbud) [inline src/modules/oauth/oauth.middleware.ts:46]: can we not define any interface or type for getUserByFirebaseId ?

File: src/modules/oauth/oauth.middleware.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+      // Set the user object on the request object
+      req.user = user;
+
+      // Check if user has the required rights
+      const userRights: any = roleRights.get(user.role);
```
Comment (sonalidevbud) [inline src/modules/oauth/oauth.middleware.ts:52]: Also let's try separating isAuthenticated and isAuthorised 2 different middleware functions

Comment (vsraccubits) [reply src/modules/oauth/oauth.middleware.ts:52]: Change completed

File: src/modules/oauth/oauth.middleware.ts (hunk 4 above)
Comment (vsraccubits) [reply src/modules/oauth/oauth.middleware.ts:46]: Removing null part from service function resolve eslint issue

image

File: src/modules/oauth/oauth.middleware.ts (hunk 3 above)
Comment (vsraccubits) [reply src/modules/oauth/oauth.middleware.ts]: Middle-ware renamed to isAuthenticated

File: src/config/config.ts (hunk 2 above)
Comment (aish-26-2000) [reply src/config/config.ts]: done

File: src/modules/oauth/oauth.service.ts (hunk 1 above)
Comment (aish-26-2000) [reply src/modules/oauth/oauth.service.ts]: done

**Comment by vsraccubits** [comment]
image

### Test case completed

**Pull request merged**